pyproject.toml

```diff
@@ -1,3 +1,266 @@
+    "collections-extended >= 2.0.2",
+
+    # HTTP server and client
+    # Latest version is necessary to work with latest werkzeug:
+    "klein >= 23.5.0",
+    # 2.2.0 has a bug: https://github.com/pallets/werkzeug/issues/2465
+    "werkzeug != 2.2.0",
+    "treq",
+    # 5.6.0 excluded because https://github.com/agronholm/cbor2/issues/208
+    "cbor2 != 5.6.0",
+
+    # 0.6 adds the ability to decode CBOR. 0.6.1 fixes PyPy.
+    "pycddl >= 0.6.1",
+
+    # Command-line parsing
+    "click >= 8.1.1",
+
+    # for pid-file support
+    "psutil",
+    "filelock",
+
+    # Duplicate the Twisted pywin32 dependency here.  See
+    # https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2392 for some
+    # discussion.
+    "pywin32 != 226;sys_platform=='win32'"
+]
+
+[project.optional-dependencies]
+tor = [
+    # 23.5 added support for custom TLS contexts in web_agent(), which is
+    # needed for the HTTP storage client to run over Tor.
+    "txtorcon >= 23.5.0",
+]
+i2p = [
+    # txi2p has Python 3 support in master branch, but it has not been
+    # released -- see https://github.com/str4d/txi2p/issues/10.  We
+    # could use a fork for Python 3 until txi2p's maintainers are back
+    # in action.  For Python 2, we could continue using the txi2p
+    # version about which no one has complained to us so far.
+    "txi2p; python_version < '3.0'",
+    "txi2p-tahoe >= 0.3.5; python_version > '3.0'",
+]
+build = [
+  "dulwich",
+  "gpg",
+  "hatchling",
+  "hatch-vcs"
+]
+
+testenv = [
+    # Pin all of these versions for the same reason you ever want to
+    # pin anything: to prevent new releases with regressions from
+    # introducing spurious failures into CI runs for whatever
+    # development work is happening at the time.  The versions
+    # selected here are just the current versions at the time.
+    # Bumping them to keep up with future releases is fine as long
+    # as those releases are known to actually work.
+    "pip==23.3.1",
+    "wheel==0.41.3",
+    "subunitreporter==23.8.0",
+    "python-subunit==1.4.2",
+    "junitxml==0.7",
+    "coverage==7.2.5",
+]
+
+# Here are the library dependencies of the test suite.
+test = [
+    "mock",
+    "pytest",
+    "pytest-twisted",
+    "hypothesis >= 3.6.1",
+    "towncrier",
+    "testtools",
+    "fixtures",
+    "beautifulsoup4",
+    "html5lib",
+    # Pin old version until
+    # https://github.com/paramiko/paramiko/issues/1961 is fixed.
+    "paramiko < 2.9",
+    "pytest-timeout",
+    # Does our OpenMetrics endpoint adhere to the spec:
+    "prometheus-client == 0.11.0",
+
+    "tahoe-lafs[tor]",  # our own "tor" extra
+    "tahoe-lafs[i2p]"  # our own "i2p" extra
+]
+
+
+[project.scripts]
+tahoe = "allmydata.scripts.runner:run"
+grid-manager = "allmydata.cli.grid_manager:grid_manager"
+
+
+[project.urls]
+Homepage = "https://tahoe-lafs.org/"
+Documentation = "https://tahoe-lafs.readthedocs.org/"
+"Source code" = "https://github.com/tahoe-lafs/tahoe-lafs/"
```

Comment:
I'm just nitpicking, but: it might be a good idea to keep all the `[project.*]` tables together under `[project]`.

Reply:
I've not done a lot of TOML: do you mean that we can do `[project]` and then have like `[urls]` inside that ...? Anyway, that sounds good if true :)

Reply:
Oh no, that is not what I meant. It would be nice if `[project]` and `[project.scripts]` and `[project.urls]` appear close to each other, as a matter of aesthetics... or convenience... or something like that. It is just a weak opinion. :-)